websocket.go

```diff
@@ -0,0 +1,277 @@
+	conn, _, err := websocket.Dial(context.Background(), p.upstreamURL, nil)
+	if err != nil {
+		return err
+	}
+	p.upstream = conn
+	return nil
+}
+
+func (p *WebSocketProxy) HandleWebSocket(w http.ResponseWriter, r *http.Request) {
+	conn, err := websocket.Accept(w, r, nil)
+	if err != nil {
+		log.Printf("WebSocket upgrade error: %v", err)
+		return
+	}
+	defer conn.CloseNow()
+
+	// Set a read timeout (e.g., 30 seconds)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel() // Ensure the context is canceled on all paths
+
+	p.mu.Lock()
+	p.clients[conn] = make(map[string]bool) // Initialize swap IDs for this client
+	p.mu.Unlock()
+
+	defer func() {
+		p.mu.Lock()
+		defer p.mu.Unlock()
+
+		// Remove client from all subscriptions
+		for swapID := range p.clients[conn] {
+			delete(p.subscribers[swapID], conn)
+			if len(p.subscribers[swapID]) == 0 {
+				// If no more clients are subscribed to this swap ID, remove it from the map
+				delete(p.subscribers, swapID)
+
+				// Optionally, send an unsubscribe message to the upstream server
+				if p.upstream != nil {
+					unsubscribeMsg := map[string]any{
+						"op":      "unsubscribe",
+						"channel": "swap.update",
+						"args":    []string{swapID},
+					}
+					messageBytes, err := json.Marshal(unsubscribeMsg)
+					if err != nil {
+						log.Printf("Failed to marshal unsubscribe message: %v", err)
+						continue
+					}
+					err = p.upstream.Write(context.Background(), websocket.MessageText, messageBytes)
+					if err != nil {
+						log.Printf("Failed to forward unsubscribe message to upstream: %v", err)
+					}
+				}
+			}
+		}
+
+		// Remove the client from the clients map
+		delete(p.clients, conn)
+	}()
+
+	for {
+		// Read message from client with timeout
+		_, message, err := conn.Read(ctx)
+		if err != nil {
+			if websocket.CloseStatus(err) == websocket.StatusNormalClosure {
+				log.Printf("Client closed the connection: %v", err)
+			} else {
+				log.Printf("WebSocket read error: %v", err)
+			}
+			break
+		}
+
+		// Parse the message
+		var msg map[string]any
+		if err := json.Unmarshal(message, &msg); err != nil {
+			log.Printf("Failed to parse client message: %v", err)
+			continue
+		}
+
+		// Handle subscription/unsubscription
+		switch msg["op"] {
+		case "subscribe":
+			p.handleSubscribe(conn, msg)
+		case "unsubscribe":
+			p.handleUnsubscribe(conn, msg)
+		default:
+			log.Printf("Unknown operation: %s", msg["op"])
+		}
+
+		// Reset the timeout for the next read
+		cancel() // Cancel the previous context
+		ctx, cancel = context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel() // Ensure the new context is canceled on all paths
+	}
+}
+
+func (p *WebSocketProxy) handleSubscribe(conn *websocket.Conn, msg map[string]any) {
+	swapIDs, ok := msg["args"].([]any)
+	if !ok {
+		log.Printf("Invalid subscribe message: %v", msg)
+		return
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
```

Comment:
This unlock can go below the for loop, before the p.upstream != nil, so the lock is not taken during the upstream write.